### PR TITLE
SCUMM: GUI: Implement original GUI and menus for v4-5-6-7 games

### PR DESCRIPTION
**Pull request by AndywinXp**
Hi everyone! Continuing my GUI saga (see #4033), here's my PR for v4-5-6-7 games.

These changes implement the GUI and menus for the full and demo* versions of:
* Passport to Adventure;
* LOOM (SCUMM v4 VGA version);
* The Secret of Monkey Island (SCUMM v4 and v5 versions);
* Monkey Island 2: LeChuck's Revenge;
* Indiana Jones and the Fate of Atlantis;
* Day of the Tentacle;
* Sam & Max: Hit the Road;
* Full Throttle;
* The Dig.

_*some demos might not allow for a menu to appear at all._

Here are some screenshots of the changes in action: https://imgur.com/a/R6tINDB

**Other info**:
* Saving and loading games is totally interchangeable with the GMM; in order to ensure a tidy experience, the thumbnail of games saved from the original menu is obtained by taking a screengrab just before the menu is drawn. Just to clarify: this doesn't save two thumbnails like COMI, it just creates it at a different point in time.
* Of course, text speed and volumes are also synced with the GMM.
* There are a lot of versions for the interpreters and their menus. As a rule of thumb I tried to always implement the most recent ones for each game (e.g. this means that v6 games have volume sliders on the menu);
* You might notice that some strings *might* be in english in some localized games. This is expected: when present, those strings are hardcoded in english in the executable in the first place, and not fetched from the game scripts.
* FM-Towns games gave me a very hard time... their menu is disabled for this PR, but the GUI banners work fine. I'll do my best to properly implement those menus in the next PR. 
* Amiga versions appear to work fine, but I don't have Mac games or MI1 SEGA CD; if anyone wants to test those, it would be great!
* The music, voice, sfx volume controls originally are mapped to keys like `{`, `}`, `[`, `]`, `\` and `;`. I don't need to tell you that would have gone south very fast had I implemented it this way, with all the different keyboards layouts. 😛 So I mapped them to `SHIFT-O/SHIFT-P`, `SHIFT-K/SHIFT-L` and `SHIFT-N/SHIFT-M`, while the text speed buttons remain the same old `+` and `-`.

That's all, I think.

Ciao!

### CODE REVIEW

**Comment by athrxx** [comment]
You were much faster than you said. So I think I have now messed up your dialogs code a bit :-(
I hope it isn't too bad, though

**Comment by AndywinXp** [comment]
> You were much faster than you said. So I think I have now messed up your dialogs code a bit :-( I hope it isn't too bad, though

No problem at all, I'll fix it tomorrow, thanks 😄

**Comment by dwatteau** [comment]
Wow, amazing and lovely work!! I'll test this with what I have there, including some Macintosh releases :) 

I see that commit 8b2855cbe0c922a2b521087f3a76b4467da3a79a contains some `sprintf_s -> sprintf` fixes for non-Windows systems. Could this particular change be squashed into the earlier commit eea3965fd7197aec83fe83cb5bc7988a262ff941 which introduced them, instead? So that we won't hit a build failure on most systems if a `git bisect` happens there. Thanks!

**Comment by AndywinXp** [comment]
> Wow, amazing and lovely work!! I'll test this with what I have there, including some Macintosh releases :)
> 
> I see that commit [8b2855c](https://github.com/scummvm/scummvm/commit/8b2855cbe0c922a2b521087f3a76b4467da3a79a) contains some `sprintf_s -> sprintf` fixes for non-C11 systems. Could this particular change be squashed into the earlier commit [eea3965](https://github.com/scummvm/scummvm/commit/eea3965fd7197aec83fe83cb5bc7988a262ff941) which introduced them, instead? So that we won't hit a build failure on most systems if a `git bisect` happens there. Thanks!

Thanks, I'll try to do that, unfortunately my git skillset is very lacking in this very precise regard 🤕

**Comment by dwatteau** [comment]
> Thanks, I'll try to do that, unfortunately my git skillset is very lacking in this very precise regard 🤕

Ah yes, Git's UX is not that great… :/ 

I'd say, as long you don't have any unstaged files or commits:

```powershell
git stash
git rebase -i eea3965~1
[go to the eea3965 commit line, change the first word in front of it to "edit", quit the rebase text editor]
[do the sprintf_s replacement again]
git add engines/scumm/gfx_gui.cpp
git commit --amend
git rebase --continue # hopefully, Git manages to integrate the change without any problem
git push --force your-upstream
git stash pop
```

but I don't want to break your tree if you're not used to doing this with Git, though

**Comment by dwatteau** [comment]
Marvellous work!

Here's what I could test and where I didn't see any issue:

* Monkey1 CD/DOS/English
* Monkey1 CD/DOS/French
* Monkey2 Amiga/English
* Monkey2 DOS/English
* Monkey2 DOS/French
* Monkey2 Macintosh/English
* Monkey2 Ultimate Talkie/English
* Indy4 CD/DOS/English
* Indy4 Floppy/DOS/French
* DOTT CD/DOS/English
* DOTT CD/DOS/French
* Full Throttle DOS/French
* Full Throttle Macintosh/French
* The Dig DOS/French
* The Dig Macintosh/French

and some versions where I saw some issues:

* Monkey2 Demo/DOS/English: SPACE shows a black bar and the menus are glitched. But that's the demo which is not supposed to be interactive. Maybe the original GUI option should just not be available for `monkey2-demo`?

* Indy4 FM-TOWNS/Japanese demo: I got it to crash in `getStringWidth()` while pressing F8. Demo is [here](https://downloads.scummvm.org/frs/demos/scumm/atlantis-fmtowns-ni-demo-jp.zip).

* Sam & Max CD/DOS/English or French: if I load an older savegame where I was holding an object, it has a permanent blue background around it

scummvm-samnmax-00000

* COMI Windows/English and French: ok from a new game, but if I load a savegame from 2 months ago, the integrated UI text becomes black

Also, if I try to call the Alt-S shortcut in order to take a screenshot within ScummVM, this unpauses the game. So I can't take a screenshot of the pause menu.

**Comment by AndywinXp** [comment]
Thanks for the thorough tests! ❤️ Glad to see that most of the games are working fine, I'll work on fixing every issue you mentioned!

**Comment by AndywinXp** [comment]
There's a caveat though: the COMI issue is known and it comes from the fact that if you load the game from the GMM, the SCUMM engine won't start the boot script, which contains the initializations of the banner colors. This means I can't do anything outside of the script scope; this is not as bad as it looks though: just reopen the main menu or change the room and it fixes itself 😄

**Comment by dwatteau** [comment]
Happy to test nice stuff like that! Ignore my comment above about the Japanese Indy4, though; I think that's just because my local build can't find the `encoding.dat` file for CJK.

**Comment by athrxx** [comment]
Maybe the Samnmax cursor can be fixed with the same postload processing that I used for EGA/VGA changes. But that's just an idea. I am not at home to actually check anything. So I don't even know if it glitches for the same reason...

**Comment by AndywinXp** [comment]
> Maybe the Samnmax cursor can be fixed with the same postload processing that I used for EGA/VGA changes. But that's just an idea. I am not at home to actually check anything. So I don't even know if it glitches for the same reason...

I might have a precise idea about it: I have some kind of cursor substitution in place both when the menu is opened, and when it is closed. It might be that I messed up the transparency color for this edge case.

File: engines/scumm/scumm-md5.h [hunk 1]
```diff
@@ -186,7 +186,7 @@ static const MD5Table md5table[] = {
 	{ "37f56ceb13e401a7ac7d9e6b37fecaf7", "loom", "EGA", "EGA", 5748, Common::EN_ANY, Common::kPlatformDOS },
 	{ "37ff1b308999c4cca7319edfcc1280a0", "puttputt", "HE 70", "Demo", 8269, Common::EN_ANY, Common::kPlatformWindows },
 	{ "3824e60cdf639d22f6df92a03dc4b131", "fbear", "HE 62", "", 7732, Common::EN_ANY, Common::kPlatformDOS },
-	{ "387a544b8b10b26912d8413bab63a853", "monkey2", "", "Demo", 9770, Common::EN_ANY, Common::kPlatformDOS },
+	{ "387a544b8b10b26912d8413bab63a853", "monkey2", "Demo", "Demo", 9770, Common::EN_ANY, Common::kPlatformDOS },
```
Comment (dwatteau) [inline engines/scumm/scumm-md5.h:189]: @AndywinXp I believe that `devtools/scumm-md5.txt` should always be updated first, or your scumm-md5.h change will be lost the next time someone runs the `devtools/md5table` tool again.

**Comment by athrxx** [comment]
Very nice! Seems to work well most of the time.

In DOTT and MI2, it seems to be rather easy to have an "convertMessageToString: buffer overflow!" error.
I just load a savegame, press F5 and then click on "Quit", press "N", click on "Quit", press "N"...
After 3 - 4 times it usually produces the error, but it is not always the same. Sometimes the error appears, when you click "Play" afterwards...

**Comment by athrxx** [comment]
Okay, that would have been too good, if that worked right out of the box...

**Comment by athrxx** [comment]
Yes, it works, it just has kind of the same layout glitches due to the font size.

**Comment by AndywinXp** [comment]
Nice! Most of it seems to be spaced as it should (do you have a picture of the original to compare?) but I'm a bit worried as why the buttons on the right are rendered that way 🤔 I'll see what I can do...

**Comment by athrxx** [comment]
Weird, the okay button here... I guess, I should fire up DOSBox and see how it looks... ;-)

**Comment by athrxx** [comment]
Looks like we're VERY close already...

**Comment by AndywinXp** [comment]
Great! I have an idea about the menu buttons:
is this piece of code even executing?
```c++
if ((_game.id == GID_DIG && _useCJKMode) &&
    (((byte)ctrl->label[0] >= 128 && (byte)ctrl->label[0] <= 159) || ((byte)ctrl->label[0] >= 224 && (byte)ctrl->label[0] <= 253))) {
     yOffset = 16;
}
```
(in `ScummEngine::drawInternalGUIControl()`)

**Comment by athrxx** [comment]
No, it never triggers.
The string seems to be the non-translated form, it start with e. g. `/BOOT.008/Save`

**Comment by AndywinXp** [comment]
> No, it never triggers. The string seems to be the non-translated form, it start with e. g. `/BOOT.008/Save`

Hmm... so it seems to need the translated string, even though the disasm doesn't seem to do that...
What happens if you try to temporarily translate the label just for that check?

**Comment by athrxx** [comment]
So, when I add a convertMessageToString() it looks a bit more promising, but still fails.
The condition `(((byte)transBuf[0] >= 128 && (byte)transBuf[0] <= 159) ||
					 ((byte)transBuf[0] >= 224 && (byte)transBuf[0] <= 253))) `
is not met. It also makes no sense for Chinese. This seems to be a check for Japanese SJIS.
Why not just use the is2ByteCharacter() method? Which will work for Japanese, Korean and Chinese...

**Comment by AndywinXp** [comment]
> So, when I add a convertMessageToString() it looks a bit more promising, but still fails. The condition `(((byte)transBuf[0] >= 128 && (byte)transBuf[0] <= 159) || ((byte)transBuf[0] >= 224 && (byte)transBuf[0] <= 253))) ` is not met. It also makes no sense for Chinese. This seems to be a check for Japanese SJIS. Why not just use the is2ByteCharacter() method? Which will work for Japanese, Korean and Chinese...

Okay, if you're sure that it `is2ByteCharacter()` is the equivalent check for that one, I will use it. Please let me know if it works as intended, so I can commit it.

**Comment by athrxx** [comment]
There is a function in the CJK DIG interpreter hat calls like dozens of times getStringHeight() on a string "ABC €x °x". Is that the menu? That's kind of what happens there, it seems to measure the height of all the strings...

But let's see if it works with the char 0 check....

**Comment by athrxx** [comment]
Okay, yes, seems to be fine! Load screen is fine, too.

**Comment by athrxx** [comment]
Korean is a bit odd... I'll have to look at the original,too...

**Comment by AndywinXp** [comment]
Yay!

> There is a function in the CJK DIG interpreter hat calls like dozens of times getStringHeight() on a string "ABC €x °x". Is that the menu? That's kind of what happens there, it seems to measure the height of all the strings...

Those calls for that particular string are a characteristic of later interpreter versions (the one akin to the COMI demo); in COMI demo they are called in showBannerAndPause() and setUpMainMenuControls().

**Comment by athrxx** [comment]
Okay, Korean does look a bit better in DOSBox.

**Comment by AndywinXp** [comment]
Oof. Is it possible that the offset correction is only for Chinese and Japanese?

**Comment by athrxx** [comment]
Hmm, I just noticed, when I looked again the Chinese DOSBox shot above that they made the slider on the bottom extra fat.

**Comment by AndywinXp** [comment]
And the sliders go a pixel higher than they do in ScummVM, while the buttons on the right seem to have a padding of one pixel on top and on the bottom in DOSBox... Jeez, and I thought we'd have CJK support right out of the box 🤣

**Comment by athrxx** [comment]
IIRC the normal interpreter does have some CJK things, but not all the things that the real CJK interpreter has.

So if you have been using that it might just be slightly different code.

**Comment by AndywinXp** [comment]
(Also, if you'd like to, please don't hesitate to commit changes regarding CJK support here, it goes without saying that I completely trust you with that)

**Comment by athrxx** [comment]
There aren't 3 different CJK interpreter versions, it is always the exact same interpreter. There is almost no difference between the way the 3 languages are treated.

**Comment by athrxx** [comment]
> (Also, if you'd like to, please don't hesitate to commit changes regarding CJK support here, it goes without saying that I completely trust you with that)

Yes, I think, for now we can be happy with what we have :-)

**Comment by AndywinXp** [comment]
> > (Also, if you'd like to, please don't hesitate to commit changes regarding CJK support here, it goes without saying that I completely trust you with that)
> 
> Yes, I think, for now we can be happy with what we have :-)

Great! 😄 moving on then...

**Comment by AndywinXp** [comment]
About the cursor transparency in v6... why on Earth did I even set 0x01 as the transparent color in `ScummEngine_v6::setCursorTransparency()` when it already was and should still be 0xFF? Maybe I was using it for v7 at some point... anyway, it's fixed! 👍

**Comment by AndywinXp** [comment]
> Very nice! Seems to work well most of the time.
> 
> In DOTT and MI2, it seems to be rather easy to have an "convertMessageToString: buffer overflow!" error. I just load a savegame, press F5 and then click on "Quit", press "N", click on "Quit", press "N"... After 3 - 4 times it usually produces the error, but it is not always the same. Sometimes the error appears, when you click "Play" afterwards...

I haven't been able to reproduce this :( I have tested MI2 Italian and English, and DOTT Italian, spamming the Quit button and N like crazy and sometime pressing Play. Maybe I'm doing it wrong? I had buffer overflow issues on convertMessageToString() while working on this, so wherever this particular crash is it's better to find it.

**Comment by AndywinXp** [comment]
Just saw that v4 uses the same menu system used here so... update incoming.

**Comment by AndywinXp** [comment]
Unless I forgot something important, I'd say that the PR has entered testing phase.
This means that every feature I wanted to add is there, and every bug that until now has been reported to me has been fixed.

(Please help me test the games if you can, there are slightly too many of them 😛)

Here are a few v4 screenshots:

image

image

image

**Comment by athrxx** [comment]
Here is another v4 screenshot... :-P

**Comment by athrxx** [comment]
I think, Hercules mode is accurate...

**Comment by AndywinXp** [comment]
>  Here is another v4 screenshot... :-P

😢.

Nice find! I originally verified the v4 GUI color values by attaching the disasm to a DOSBox-X debugger, so this suggests me that they outright change for the CGA mode; I'll see what I can do...

**Comment by athrxx** [comment]
Yes, that would be my guess, too. Hercules seems to be totally fine, although it has the more elaborate dithering pattern. So it is probably a CGA specific change of colors...

**Comment by AndywinXp** [comment]
Yes, the color array was indeed changed with this render mode; I think I fixed it:
image

This should be applied to both CGA and CGAComp mode, is it right?

**Comment by athrxx** [comment]
Yes, CGA composite is only supported in MM/ZAK v1. So just CGA is fine...

**Comment by AndywinXp** [comment]
Yesterday I finished my (first, actually!) full playthrough of Indy4 (CD Talkie), saving at every occasion I could, and it went very smoothly. The only issue I encountered regarding this PR is that I didn't restore the shake effect when exiting a menu, so I fixed that.

**Comment by athrxx** [comment]
The PR should now be ready.
@AndywinXp has tackled all issues that we found.
So, if there aren't any objections we're going to merge this in a couple of days.

**Comment by orgads** [comment]
Gong?

**Comment by athrxx** [comment]
Okay, let's do it...

**Pull request merged**